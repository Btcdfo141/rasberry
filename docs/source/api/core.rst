.. _core_module:

:mod:`spencerassistant.core`
-------------------------

.. automodule:: spencerassistant.core
    :members: