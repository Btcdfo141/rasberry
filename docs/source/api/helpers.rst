:mod:`spencerassistant.helpers`
============================

.. automodule:: spencerassistant.helpers
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.aiohttp\_client
-------------------------------------

.. automodule:: spencerassistant.helpers.aiohttp_client
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.area\_registry
------------------------------------

.. automodule:: spencerassistant.helpers.area_registry
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.check\_config
-----------------------------------

.. automodule:: spencerassistant.helpers.check_config
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.collection
--------------------------------

.. automodule:: spencerassistant.helpers.collection
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.condition
-------------------------------

.. automodule:: spencerassistant.helpers.condition
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.config\_entry\_flow
-----------------------------------------

.. automodule:: spencerassistant.helpers.config_entry_flow
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.config\_entry\_oauth2\_flow
-------------------------------------------------

.. automodule:: spencerassistant.helpers.config_entry_oauth2_flow
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.config\_validation
----------------------------------------

.. automodule:: spencerassistant.helpers.config_validation
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.data\_entry\_flow
---------------------------------------

.. automodule:: spencerassistant.helpers.data_entry_flow
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.debounce
------------------------------

.. automodule:: spencerassistant.helpers.debounce
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.deprecation
---------------------------------

.. automodule:: spencerassistant.helpers.deprecation
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.device\_registry
--------------------------------------

.. automodule:: spencerassistant.helpers.device_registry
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.discovery
-------------------------------

.. automodule:: spencerassistant.helpers.discovery
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.dispatcher
--------------------------------

.. automodule:: spencerassistant.helpers.dispatcher
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.entity
----------------------------

.. automodule:: spencerassistant.helpers.entity
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.entity\_component
---------------------------------------

.. automodule:: spencerassistant.helpers.entity_component
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.entity\_platform
--------------------------------------

.. automodule:: spencerassistant.helpers.entity_platform
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.entity\_registry
--------------------------------------

.. automodule:: spencerassistant.helpers.entity_registry
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.entity\_values
------------------------------------

.. automodule:: spencerassistant.helpers.entity_values
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.entityfilter
----------------------------------

.. automodule:: spencerassistant.helpers.entityfilter
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.event
---------------------------

.. automodule:: spencerassistant.helpers.event
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.icon
--------------------------

.. automodule:: spencerassistant.helpers.icon
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.integration\_platform
-------------------------------------------

.. automodule:: spencerassistant.helpers.integration_platform
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.intent
----------------------------

.. automodule:: spencerassistant.helpers.intent
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.json
--------------------------

.. automodule:: spencerassistant.helpers.json
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.location
------------------------------

.. automodule:: spencerassistant.helpers.location
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.logging
-----------------------------

.. automodule:: spencerassistant.helpers.logging
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.network
-----------------------------

.. automodule:: spencerassistant.helpers.network
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.restore\_state
------------------------------------

.. automodule:: spencerassistant.helpers.restore_state
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.script
----------------------------

.. automodule:: spencerassistant.helpers.script
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.service
-----------------------------

.. automodule:: spencerassistant.helpers.service
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.signal
-----------------------------

.. automodule:: spencerassistant.helpers.signal
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.state
---------------------------

.. automodule:: spencerassistant.helpers.state
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.storage
-----------------------------

.. automodule:: spencerassistant.helpers.storage
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.sun
-------------------------

.. automodule:: spencerassistant.helpers.sun
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.system\_info
----------------------------------

.. automodule:: spencerassistant.helpers.system_info
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.temperature
---------------------------------

.. automodule:: spencerassistant.helpers.temperature
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.template
------------------------------

.. automodule:: spencerassistant.helpers.template
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.translation
---------------------------------

.. automodule:: spencerassistant.helpers.translation
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.typing
----------------------------

.. automodule:: spencerassistant.helpers.typing
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.helpers.update\_coordinator
-----------------------------------------

.. automodule:: spencerassistant.helpers.update_coordinator
   :members:
   :undoc-members:
   :show-inheritance:
