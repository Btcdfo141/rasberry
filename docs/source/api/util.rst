:mod:`spencerassistant.util`
=========================

.. automodule:: spencerassistant.util
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.yaml
-----------------------

.. automodule:: spencerassistant.util.yaml
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.aiohttp
--------------------------

.. automodule:: spencerassistant.util.aiohttp
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.async\_
--------------------------

.. automodule:: spencerassistant.util.async_
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.color
------------------------

.. automodule:: spencerassistant.util.color
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.decorator
----------------------------

.. automodule:: spencerassistant.util.decorator
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.distance
---------------------------

.. automodule:: spencerassistant.util.distance
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.dt
---------------------

.. automodule:: spencerassistant.util.dt
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.json
-----------------------

.. automodule:: spencerassistant.util.json
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.location
---------------------------

.. automodule:: spencerassistant.util.location
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.logging
--------------------------

.. automodule:: spencerassistant.util.logging
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.network
--------------------------

.. automodule:: spencerassistant.util.network
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.package
--------------------------

.. automodule:: spencerassistant.util.package
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.pil
----------------------

.. automodule:: spencerassistant.util.pil
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.pressure
---------------------------

.. automodule:: spencerassistant.util.pressure
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.ssl
----------------------

.. automodule:: spencerassistant.util.ssl
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.temperature
------------------------------

.. automodule:: spencerassistant.util.temperature
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.unit\_system
-------------------------------

.. automodule:: spencerassistant.util.unit_system
   :members:
   :undoc-members:
   :show-inheritance:

spencerassistant.util.volume
-------------------------

.. automodule:: spencerassistant.util.volume
   :members:
   :undoc-members:
   :show-inheritance:
