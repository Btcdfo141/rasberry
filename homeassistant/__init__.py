"""Init file for Home Assistant."""


test2
