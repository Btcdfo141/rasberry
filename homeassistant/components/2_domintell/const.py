"""Constants for the Domintell Integration integration."""

DOMAIN = "2_domintell"
