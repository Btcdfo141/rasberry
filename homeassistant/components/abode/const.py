"""Constants for the Abode Security System component."""
import logging

LOGGER = logging.getLogger(__package__)

DOMAIN = "abode"
ATTRIBUTION = "Data provided by goabode.com"

DEFAULT_CACHEDB = "jaraco_abode_cache.pickle"
CONF_POLLING = "polling"
