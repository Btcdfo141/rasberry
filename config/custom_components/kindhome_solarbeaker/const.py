from datetime import timedelta

from homeassistant.const import Platform

DOMAIN="kindhome_solarbeaker"
PLATFORMS = [Platform.COVER]

UPDATE_INTERVAL = timedelta(seconds=20)

SERVICE_UUID = "75c276c3-8f97-20bc-a143-b354244886d4"
TITLE="Kindhome Solarbeaker"

DATA_DEVICE = "device"
DATA_COOR = "coordinator"
