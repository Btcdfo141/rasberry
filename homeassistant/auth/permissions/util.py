"""Helpers to deal with permissions."""
from functools import wraps

from typing import (  # noqa: F401
    TYPE_CHECKING, Callable, Dict, List, Optional, Union, cast)

from homeassistant.exceptions import Unauthorized, UnknownUser

from .const import POLICY_CONTROL
from .models import PermissionLookup
from .types import CategoryType, SubCategoryDict, ValueType

if TYPE_CHECKING:
    from homeassistant.core import HomeAssistant, Service, ServiceCall  # noqa

LookupFunc = Callable[[PermissionLookup, SubCategoryDict, str],
                      Optional[ValueType]]
SubCatLookupType = Dict[str, LookupFunc]


def authorized_service_call(hass: 'HomeAssistant', domain: str) -> Callable:
    """Ensure user of a config entry-enabled service call has permission."""
    def decorator(service: 'Service') -> Callable:
        """Decorate."""
        @wraps(service)
        async def check_permissions(call: 'ServiceCall') -> None:
            """Check user permission and raise before call if unauthorized."""
            if not call.context.user_id:
                return

            user = await hass.auth.async_get_user(call.context.user_id)
            if user is None:
                raise UnknownUser(
                    context=call.context,
                    permission=POLICY_CONTROL
                )

            # If the user passes one or more entity IDs, check permissions
            # there; otherwise, check permissions against entities registered
            # to the domain:
            if call.data.get('entity_id'):
                if isinstance(call.data['entity'], str):
                    entities = [call.data['entity_id']]
                else:
                    entities = call.data['entity_id']
            else:
                reg = await hass.helpers.entity_registry.async_get_registry()
                entities = [
                    entity.entity_id for entity in reg.entities.values()
                    if entity.platform == domain
                ]

            for entity_id in entities:
                if user.permissions.check_entity(entity_id, POLICY_CONTROL):
                    return await service(call)

            raise Unauthorized(
                context=call.context,
                permission=POLICY_CONTROL,
            )
        return check_permissions
    return decorator


def lookup_all(perm_lookup: PermissionLookup, lookup_dict: SubCategoryDict,
               object_id: str) -> ValueType:
    """Look up permission for all."""
    # In case of ALL category, lookup_dict IS the schema.
    return cast(ValueType, lookup_dict)


def compile_policy(
        policy: CategoryType, subcategories: SubCatLookupType,
        perm_lookup: PermissionLookup
    ) -> Callable[[str, str], bool]:  # noqa
    """Compile policy into a function that tests policy.
    Subcategories are mapping key -> lookup function, ordered by highest
    priority first.
    """
    # None, False, empty dict
    if not policy:
        def apply_policy_deny_all(entity_id: str, key: str) -> bool:
            """Decline all."""
            return False

        return apply_policy_deny_all

    if policy is True:
        def apply_policy_allow_all(entity_id: str, key: str) -> bool:
            """Approve all."""
            return True

        return apply_policy_allow_all

    assert isinstance(policy, dict)

    funcs = []  # type: List[Callable[[str, str], Union[None, bool]]]

    for key, lookup_func in subcategories.items():
        lookup_value = policy.get(key)

        # If any lookup value is `True`, it will always be positive
        if isinstance(lookup_value, bool):
            return lambda object_id, key: True

        if lookup_value is not None:
            funcs.append(_gen_dict_test_func(
                perm_lookup, lookup_func, lookup_value))

    if len(funcs) == 1:
        func = funcs[0]

        @wraps(func)
        def apply_policy_func(object_id: str, key: str) -> bool:
            """Apply a single policy function."""
            return func(object_id, key) is True

        return apply_policy_func

    def apply_policy_funcs(object_id: str, key: str) -> bool:
        """Apply several policy functions."""
        for func in funcs:
            result = func(object_id, key)
            if result is not None:
                return result
        return False

    return apply_policy_funcs


def _gen_dict_test_func(
        perm_lookup: PermissionLookup,
        lookup_func: LookupFunc,
        lookup_dict: SubCategoryDict
    ) -> Callable[[str, str], Optional[bool]]:  # noqa
    """Generate a lookup function."""
    def test_value(object_id: str, key: str) -> Optional[bool]:
        """Test if permission is allowed based on the keys."""
        schema = lookup_func(
            perm_lookup, lookup_dict, object_id)  # type: ValueType

        if schema is None or isinstance(schema, bool):
            return schema

        assert isinstance(schema, dict)

        return schema.get(key)

    return test_value
