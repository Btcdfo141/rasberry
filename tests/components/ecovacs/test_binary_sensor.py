"""Tests for Ecovacs binary sensors."""

from deebot_client.event_bus import EventBus
from deebot_client.events import WaterAmount, WaterInfoEvent
import pytest
from syrupy import SnapshotAssertion

from homeassistant.components.ecovacs.controller import EcovacsController
from homeassistant.const import STATE_OFF, STATE_UNKNOWN
from homeassistant.core import HomeAssistant
from homeassistant.helpers import device_registry as dr, entity_registry as er

from .util import notify_and_wait

pytestmark = [pytest.mark.usefixtures("init_integration")]


async def test_mop_attached(
    hass: HomeAssistant,
    device_registry: dr.DeviceRegistry,
    entity_registry: er.EntityRegistry,
    controller: EcovacsController,
    snapshot: SnapshotAssertion,
) -> None:
    """Test mop_attached binary sensor."""
    assert (state := hass.states.get("binary_sensor.ozmo_950_mop_attached"))
    assert state.state == STATE_UNKNOWN

    assert (entity_entry := entity_registry.async_get(state.entity_id))
    assert snapshot == entity_entry

    assert entity_entry.device_id
    assert (device_entry := device_registry.async_get(entity_entry.device_id))
    assert snapshot == device_entry

    event_bus: EventBus = controller.devices[0].events
    await notify_and_wait(
        hass, event_bus, WaterInfoEvent(WaterAmount.HIGH, mop_attached=True)
    )

    assert (state := hass.states.get(state.entity_id))
    assert snapshot == state

    await notify_and_wait(
        hass, event_bus, WaterInfoEvent(WaterAmount.HIGH, mop_attached=False)
    )

    assert (state := hass.states.get(state.entity_id))
    assert state.state == STATE_OFF
