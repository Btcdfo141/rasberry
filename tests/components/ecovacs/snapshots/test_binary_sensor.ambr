# serializer version: 1
# name: test_mop_attached
  EntityRegistryEntrySnapshot({
    'aliases': set({
    }),
    'area_id': None,
    'capabilities': None,
    'config_entry_id': <ANY>,
    'device_class': None,
    'device_id': <ANY>,
    'disabled_by': None,
    'domain': 'binary_sensor',
    'entity_category': <EntityCategory.DIAGNOSTIC: 'diagnostic'>,
    'entity_id': 'binary_sensor.ozmo_950_mop_attached',
    'has_entity_name': True,
    'hidden_by': None,
    'icon': None,
    'id': <ANY>,
    'name': None,
    'options': dict({
    }),
    'original_device_class': None,
    'original_icon': None,
    'original_name': 'Mop attached',
    'platform': 'ecovacs',
    'previous_unique_id': None,
    'supported_features': 0,
    'translation_key': 'mop_attached',
    'unique_id': 'E1234567890000000001_mop_attached',
    'unit_of_measurement': None,
  })
# ---
# name: test_mop_attached.1
  DeviceRegistryEntrySnapshot({
    'area_id': None,
    'config_entries': <ANY>,
    'configuration_url': None,
    'connections': set({
    }),
    'disabled_by': None,
    'entry_type': None,
    'hw_version': None,
    'id': <ANY>,
    'identifiers': set({
      tuple(
        'ecovacs',
        'E1234567890000000001',
      ),
    }),
    'is_new': False,
    'manufacturer': 'Ecovacs',
    'model': 'DEEBOT OZMO 950 Series',
    'name': 'Ozmo 950',
    'name_by_user': None,
    'serial_number': 'E1234567890000000001',
    'suggested_area': None,
    'sw_version': None,
    'via_device_id': None,
  })
# ---
# name: test_mop_attached.2
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'friendly_name': 'Ozmo 950 Mop attached',
    }),
    'context': <ANY>,
    'entity_id': 'binary_sensor.ozmo_950_mop_attached',
    'last_changed': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
